server/server.go

```diff
@@ -0,0 +1,244 @@
+	if s.simulation {
+		s.simulationCancel()
+	}
+	return s.s.Shutdown(ctx)
+}
+
+func (s *Server) handleReady(w http.ResponseWriter, r *http.Request) {
+	start := time.Now()
+	ws := newResponseWriterWithStatus(w)
+	defer s.reportLatency(start, ws.status, r.Method, "ready")
+	discardBody(r)
+	switch r.Method {
+	case http.MethodGet:
+		ws.WriteHeader(http.StatusOK)
+	default:
+		http.Error(w, "", http.StatusNotFound)
+	}
+}
+
+func (s *Server) handleMhSubtree(w http.ResponseWriter, r *http.Request) {
+	start := time.Now()
+	switch r.Method {
+	case http.MethodGet:
+		if s.simulation {
+			s.simulationJobs <- r
+			http.Error(w, "", http.StatusNotFound)
+		} else {
+			ws := newResponseWriterWithStatus(w)
+			defer s.reportLatency(start, ws.status, r.Method, "multihash")
+			s.handleGetMh(newIPNILookupResponseWriter(ws, preferJSON), r)
+		}
+	default:
+		discardBody(r)
+		http.Error(w, "", http.StatusNotFound)
+	}
+}
+
+func (s *Server) handleGetMh(w lookupResponseWriter, r *http.Request) {
+	if err := w.Accept(r); err != nil {
+		switch e := err.(type) {
+		case errHttpResponse:
+			e.WriteTo(w)
+		default:
+			logger.Errorw("Failed to accept lookup request", "err", err)
+			http.Error(w, "", http.StatusInternalServerError)
+		}
+		return
+	}
+	ctx := context.Background()
+
+	c, err := finderhttpclient.NewDHashClient(s.dhaddr, s.dhaddr)
+	if err != nil {
+		s.handleError(w, err)
+		return
+	}
+
+	findResponse, err := c.Find(ctx, w.Key())
+
+	if err != nil {
+		s.handleError(w, err)
+		return
+	}
+
+	// There going to be exactly one item in the array as we seached for one multihash only
+	// and if it wasn't found then an error would have been returned
+	mhr := findResponse.MultihashResults[0]
+
+	for _, pr := range mhr.ProviderResults {
+		if err := w.WriteProviderResult(pr); err != nil {
+			logger.Errorw("Failed to encode provider result", "err", err)
+			http.Error(w, "", http.StatusInternalServerError)
+			return
+		}
+	}
+	if err := w.Close(); err != nil {
+		switch e := err.(type) {
+		case errHttpResponse:
+			e.WriteTo(w)
+		default:
+			logger.Errorw("Failed to finalize lookup results", "err", err)
+			http.Error(w, "", http.StatusInternalServerError)
+		}
+	}
+}
+
+func (s *Server) reportLatency(start time.Time, status int, method, path string) {
+	s.m.RecordHttpLatency(context.Background(), time.Since(start), method, path, status)
+}
+
+func discardBody(r *http.Request) {
+	_, _ = io.Copy(io.Discard, r.Body)
+	_ = r.Body.Close()
+}
+
+func (s *Server) handleCatchAll(w http.ResponseWriter, r *http.Request) {
+	discardBody(r)
+	http.Error(w, "", http.StatusNotFound)
+}
+
+func (s *Server) handleError(w http.ResponseWriter, err error) {
+	var status int
+	switch err.(type) {
+	case ErrUnsupportedMulticodecCode, ErrMultihashDecode:
+		status = http.StatusBadRequest
+	default:
+		status = http.StatusInternalServerError
+	}
+	http.Error(w, err.Error(), status)
+}
+
+func (s *Server) simulationWorker() {
+	select {
```

Comment:
Missing `for {}` loop around select, otherwise it'll only handle a single job ish.

Reply:
Great spot, adding